File: 3_linkedList/LinkedList.ts
/* eslint-disable @typescript-eslint/no-non-null-assertion */

import assert from 'assert'

import { LinkedListNode } from './LinkedListNode'

/**
 * 链表实现的双端队列
 */
class LinkedList<E = number> {
  /** 哨兵 */
  readonly root: LinkedListNode<E>
  private _size = 0

  /**
   * 初始化双向链表，判断节点时 next/pre 若为 root，则表示 next/pre 为空
   */
  constructor(iterable?: Iterable<E>) {
    this.root = new LinkedListNode<E>(undefined)
    this.root.pre = this.root
    this.root.next = this.root

    for (const item of iterable || []) {
      this.push(item)
      this._size++
    }
  }

  unshift(val: E): void {
    this.root.insertAfter(new LinkedListNode(val))
    this._size++
  }

  shift(): E | undefined {
    if (this._isEmpty()) return undefined
    this._size--
    // eslint-disable-next-line prefer-destructuring
    const next = this.root.next
    return next ? next.remove().value : undefined
  }

  push(val: E): void {
    this.root.insertBefore(new LinkedListNode(val))
    this._size++
  }

  pop(): E | undefined {
    if (this._isEmpty()) return undefined
    this._size--
    // eslint-disable-next-line prefer-destructuring
    const pre = this.root.pre
    return pre ? pre.remove().value : undefined
  }

  first(): E | undefined {
    // eslint-disable-next-line prefer-destructuring
    const next = this.root.next
    return next ? next.value : undefined
  }

  last(): E | undefined {
    // eslint-disable-next-line prefer-destructuring
    const pre = this.root.pre
    return pre ? pre.value : undefined
  }

  toString(): string {
    return `${[...this]}`
  }

  insert(node: LinkedListNode<E>, cur: E): LinkedListNode<E> {
    const newNode = new LinkedListNode(cur)
    node.insertBefore(newNode)
    this._size++
    return newNode
  }

  erase(node: LinkedListNode<E>): void {
    node.remove()
    this._size--
  }

  // eslint-disable-next-line generator-star-spacing
  *[Symbol.iterator](): IterableIterator<E> {
    let node = this.root.next!
    while (node !== this.root) {
      yield node.value!
      node = node.next!
    }
  }

  get length(): number {
    return this._size
  }

  private _isEmpty(): boolean {
    return this.root.next === this.root
  }
}

if (require.main === module) {
  const nums = new LinkedList([1, 2, 3, 4, 5])
  assert.strictEqual(nums.shift(), 1)
  for (const num of nums) console.log(num)
  console.log(`${nums}`)
}

export { LinkedList }

// java查找链表元素：起点折半查找 这样最坏情况也只要找一半就可以了。
// Node<E> node(int index) {
//   assert isElementIndex(index);

//   if (index < (size >> 1)) {
//       Node<E> x = first;
//       for (int i = 0; i < index; i++)
//           x = x.next;
//       return x;
//   } else {
//       Node<E> x = last;
//       for (int i = size - 1; i > index; i--)
//           x = x.prev;
//       return x;
//   }
// }

File: 算法竞赛进阶指南/GoDS (Go Data Structures)/src/deque/linkedlist/linkedlist.go
package linkedlist

import (
	"cmnx/src/deque"
	"fmt"
)

// Assert Deque implementation
var _ deque.Deque = (*LinkedList)(nil)

func demo() {
	linkedList := NewLinkedList()
	linkedList.Append(1)
	linkedList.Append(2)
	linkedList.Append(3)
	fmt.Println(linkedList.At(0))
	linkedList.Pop()
	linkedList.ForEach(func(value interface{}, index int) {
		fmt.Println(value, index)
	})
}

func NewLinkedList() *LinkedList {
	root := NewLinkedListNode(nil, nil, nil)
	root.Pre = root
	root.Next = root
	return &LinkedList{root, 0}
}

type LinkedList struct {
	root *LinkedListNode
	size int
}

func (list *LinkedList) Append(value interface{}) {
	node := NewLinkedListNode(value, nil, nil)
	list.root.InsertBefore(node)
	list.size++
}

func (list *LinkedList) Pop() (value interface{}) {
	if list.size == 0 {
		return
	}

	value = list.root.Pre.Remove().Value
	list.size--
	return

}

func (list *LinkedList) AppendLeft(value interface{}) {
	node := NewLinkedListNode(value, nil, nil)
	list.root.InsertAfter(node)
	list.size++
}

func (list *LinkedList) PopLeft() (value interface{}) {
	if list.size == 0 {
		return
	}

	value = list.root.Next.Remove().Value
	list.size--
	return
}

func (list *LinkedList) Len() int {
	return list.size
}

func (list *LinkedList) At(index int) (value interface{}) {
	if index < 0 {
		index += list.size
	}

	if index < 0 || index >= list.size {
		return
	}

	node := list.root.Next
	for i := 0; i < index; i++ {
		node = node.Next
	}
	value = node.Value
	return
}

func (list *LinkedList) ForEach(f func(value interface{}, index int)) {
	node := list.root.Next
	for i := 0; i < list.size; i++ {
		f(node.Value, i)
		node = node.Next
	}
}

func NewLinkedListNode(value interface{}, pre, next *LinkedListNode) *LinkedListNode {
	return &LinkedListNode{
		Value: value,
	}
}

type LinkedListNode struct {
	Value interface{}
	Pre   *LinkedListNode
	Next  *LinkedListNode
}

// 在other之后插入新节点 并返回新节点
func (cur *LinkedListNode) InsertAfter(other *LinkedListNode) *LinkedListNode {
	other.Pre = cur
	other.Next = cur.Next
	other.Pre.Next = other
	if other.Next != nil {
		other.Next.Pre = other
	}
	return other
}

// 在other之前插入新节点 并返回新节点
func (cur *LinkedListNode) InsertBefore(other *LinkedListNode) *LinkedListNode {
	other.Pre = cur.Pre
	other.Next = cur
	other.Next.Pre = other
	if other.Pre != nil {
		other.Pre.Next = other
	}
	return other
}

// 从链表里移除自身
func (cur *LinkedListNode) Remove() *LinkedListNode {
	if cur.Pre != nil {
		cur.Pre.Next = cur.Next
	}
	if cur.Next != nil {
		cur.Next.Pre = cur.Pre
	}
	cur.Pre = nil
	cur.Next = nil
	return cur
}
